compiler/rustc_mir_transform/src/lib.rs

```diff
@@ -165,6 +170,20 @@ fn mir_keys(tcx: TyCtxt<'_>, (): ()) -> FxHashSet<LocalDefId> {
     set
 }
 
+/// Returns array of all the function or closures `DefId`s in this crate that have MIR
+fn mir_fn_and_closures_keys<'tcx>(tcx: TyCtxt<'tcx>, _: ()) -> &'tcx [LocalDefId] {
```

Comment:
Why do you need to introduce this other query?

Reply:
To cache it's results. 

Number of calls can change inside some function `f(x)`, but that doesn't mean the count of functions will change.

Reply:
`mir_inliner_crate_fn_callees_count_map` is a global analysis of all functions in the crate. We pretty much expect to recompute during each compilation session. `mir_inliner_crate_fn_callees_count` is already there to shield downstream code from invalidations.
If you want to avoid spurious invalidations, I'd recommend listing definitions by using `mir_keys` followed by the filter: `tcx.hir().body_owners` creates a dependency on the full HIR, so may get invalidated by span changes. `mir_keys` does not have this issue.

Reply:
@cjgillot so I should inline this function to where it is used? Did I understood correctly?